### PR TITLE
[5.6] View/Translator/Migrator configuration via ServiceProvider only after resolving (no force resolving).

### PR DESCRIPTION
**Pull request by ed-fruty**
Now, if even my application don't interacts with `view` (API for example), `view` will still be created by IoC.
This is because `PaginationServiceProvider` uses `loadViewsFrom` method, which call `view` service.

Such behavior already with `translator`. 

And any other service provider can do the same.

The main thing is do not boot services until it needs.

This PR checks, if service (view/translator/migrator) was already resolved, then setup it right now (add namespace for `view` etc.). 
But if it not loaded yet, subscribes to service resolving and apply changes, when it will be ready.

### CODE REVIEW

**Comment by browner12** [comment]
is this different than 'deferred' providers?

**Comment by ed-fruty** [comment]
@browner12 , yes it different. Service provider can be not deferred, but it does not have to initialize other services that may not be needed.

Type `php artisan tinker` and after that `app()->resolved('view')`. It will be `true`, and do you really need it, that `view` is resolved always ? and you can't control that, because any provider can load it using `loadViewsFrom`.

**Comment by taylorotwell** [comment]
In a standard Laravel application with a single database query how much of a performance impact does this make (before / after)?

**Comment by ed-fruty** [comment]
The question is not about performance degree.
We can load all defined services (queue, db etc.) force, ant it will add 0,001-0,002 sec, but the question is `why` ? if we can not do this.

let's imagine we have two group of routes in out application: `api` and `web`.
My `web` group render views and uses namespaces, view composers, view creators.
When I want to add new view composer for example, I need to do something like:

```php
class AppServiceProvider
{
   public function boot()
   {
          View::composer('my.view', Composer::class);
   }
}
```

View Composers/ creators/ namespaces can be a lot. And they applies for any of route groups (web/api).
I can't use `afterResolving` method for `view` service, because it was already resolved.

With this PR we can do lazy view configuration, like:

```php
class AppServiceProvider
{
   public function boot()
   {
       $this->app->afterResolving('view', function (Factory $view) { 
             $view->addNamespace('namespace', 'path');
             $view->composer('view', Composer::class);
       });
   }
}
```
And it will setup `view` only when needs to render something. It doesn't apply any side affect for my `api` part of application.

**Comment by taylorotwell** [comment]
All those tests you added could basically be removed I think. There is too much mocking to be useful at testing anything.

**Comment by ed-fruty** [comment]
Done.

And maybe it will be useful to update docs and change (Validator extending/View composers/ etc. ) examples from directly calling services to using `afterResolving` ?

File: src/Illuminate/Support/ServiceProvider.php [hunk 1]
```diff
@@ -179,6 +198,22 @@ protected function addPublishGroup($group, $paths)
...
 
+    /**
+     * Apply callback to service when it ready.
+     *
+     * If service resolved, apply now.
+     * If not resolved, apply after resolving.
+     *
+     * @param mixed|string|object $abstract
+     * @param callable $callback
+     */
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/ServiceProvider.php]: missing return annotation

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by taylorotwell** [comment]
I just don't see a big point in changing these things unless we have clear performance before / after comparisons.

**Comment by ed-fruty** [comment]
@freekmurze , @GrahamCampbell , @sisve 
Guys, what do you think about this?

**Pull request closed without merging**